binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+    We'll see some examples of this later in the chapter.
+
+The Apache Parquet file format ([parquet.apache.org](https://parquet.apache.org/)) is designed to address these limitations.
+Like a csv file, a parquet file stores a single rectangular data set.
+Features of the parquet format that make it appealing for data scientists include:
+
+-   Parquet files store data in a binary format (not plain text), avoiding inaccuracies in encoding and making it faster to read and write than a csv file.
+-   Parquet files store metadata that uniquely identifies the data types stored in each column, which avoids the ambiguity of csv files.
+-   Parquet files are compressed by default, and are typically much smaller than the corresponding csv files would be.
+-   Parquet files internally break the data into distinct "chunks" each with associated metadata, making it possible to write intelligent file readers that can jump straight to a specific part of the data when the full data set is not needed.
+
+All these factors together make the parquet format appealing when working with larger data sets.
+
+The arrow package supplies `read_parquet()` and `write_parquet()` functions that can read and write data in the parquet format.
+We'll use these functions to explore the advantages to parquet files.
+
+### Parquet files are unambiguous
+
+Recall this example from earlier:
+
+```{r ambiguous-data-frame}
+original <- tibble(
+  integers = 1L:3L,
+  strings = c("2022-01-01", "1980-12-01", "1900-12-12")
+)
+original
+```
+
+As we saw, writing this data to a csv file and reading it back in leads to a mistake in identifying the data types for each column.
+The parquet format avoids this ambiguity.
+When we write the data set to a parquet file using `write_parquet()` and then read it back in with `read_parquet()`, the data types are preserved:
+
+```{r read-parquet-unambigiously}
+parquet_file <- tempfile(fileext = ".parquet")
+write_parquet(original, parquet_file)
+read_parquet(parquet_file)
+```
+
+### Parquet files are binary
+
+Earlier in this section we saw an example in which csv files store numeric data inaccurately because they use plain text rather than binary format.
+Let's repeat the exercise using the binary parquet format.
+We first write some random numbers to a parquet file:
+
+```{r parquet-precision-write}
+#| message: false
+parquet_file <- tempfile(fileext = ".parquet")
+original <- tibble(old_x = rnorm(1000))
+write_parquet(original, parquet_file)
+```
+
+Next we load the file and give the reloaded variable a new name:
+
+```{r parquet-precision-read}
+reloaded <- read_parquet(parquet_file) |>
+  rename(new_x = old_x)
+```
+
+Finally we bind the old and new data sets together and compare them:
+
+```{r parquet-precision-success}
+bind_cols(original, reloaded) |> 
+  mutate(same_value = old_x == new_x) |>
+  count(same_value)
+```
+
+As you can see, numeric values stored in parquet files are stored faithfully.
```

Comment:
Perhaps worth noting (maybe above when you talk about the benefits of Parquet) that Parquet (and Arrow for that matter) support more types than R does. So you can get better precision and support operations (like calculations on decimal types) that are not possible in R. In fact, you may get a loss of fidelity reading some types into R.